### PR TITLE
refactor RangeIndexReader to make it easier to evolve its implementation

### PR DESCRIPTION
**Pull request by richardstartin**
## Description
This PR refactors the interface between the `RangeIndexBasedFilterOperator` and `RangeIndexReader`, the latter of which I plan to submit a proposal to enhance. This work is being done in advance of submitting the enhancement proposal to limit the amount of change necessary. 

The pending proposal for a new range index will support fast exact matches without the need to scan, and this PR tries to make it possible to model both exact and inexact indexing with the same interface, allowing the `RangeIndexReaderImpl` to decide the implementation to use based on the version in the index header.

I actually found 2 bugs in the current implementation doing this:
* Off by one bug in the range sizing. For 1000 distinct values, there should be 50 values in each bucket, but 51 are assigned to the first 19 buckets and only 31 to the last bucket. I left this in place.
* `float` values were converted to `int` which interferes with bucketing, meaning in the test for float all 20 buckets were collapsed into one. I fixed this. This flaw could have been caught by clearly separating preconditions (i.e. ranges) from the system under test, but slipped through because the bucketing logic was assumed as a precondition of the test.

/cc @Jackie-Jiang @mayankshriv 

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/RangeIndexCreator.java [hunk 1]
```diff
@@ -539,7 +543,7 @@ public int compare(Number val1, Number val2) {
 
     @Override
     public void put(int position, Number value) {
-      _dataBuffer.putFloat(position << 2, value.intValue());
+      _dataBuffer.putFloat(position << 2, value.floatValue());
```
Comment (richardstartin) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/RangeIndexCreator.java:546]: This fixes #7436

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/RangeIndexReader.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+ */
+package org.apache.pinot.segment.spi.index.reader;
+
+import java.io.Closeable;
+
+/**
+ * Interface for indexed range queries
+ * @param <T>
+ */
+public interface RangeIndexReader<T> extends Closeable {
```
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/RangeIndexReader.java:27]: Does it make sense to fix the type to `ImmutableBitmapDataProvider`?

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/RangeIndexReader.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  /**
+   * Returns doc ids with a value between min and max, both inclusive.
+   * @param min the inclusive lower bound.
+   * @param max the inclusive upper bound.
+   * @return the matching doc ids.
+   */
+  T getMatchingDocIds(long min, long max);
```
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/RangeIndexReader.java:34]: Annotate the methods return value as `Nullable` and add javadoc on when to return `null`

Comment (Jackie-Jiang) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/RangeIndexReader.java:34]: (nit) Re-order the methods following `int, long, float, double` to be consistent with other places for readability

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/RangeIndexReaderImpl.java [hunk 4]
```diff
@@ -217,4 +272,26 @@ public void close() {
     // NOTE: DO NOT close the PinotDataBuffer here because it is tracked by the caller and might be reused later. The
     // caller is responsible of closing the PinotDataBuffer.
   }
+
+  private ImmutableRoaringBitmap getMatchesInRange(int firstRangeId, int lastRangeId) {
+    if (firstRangeId == lastRangeId) {
+      return null;
+    }
+    MutableRoaringBitmap matching = new MutableRoaringBitmap();
+    for (int rangeId = firstRangeId + 1; rangeId < lastRangeId; ++rangeId) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/RangeIndexReaderImpl.java:281]: Won't work if `lastRangeId` is -1?

Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/RangeIndexReaderImpl.java:281]: (nit) We usually put `rangeId++`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/RangeIndexReaderImpl.java [hunk 5]
```diff
@@ -217,4 +272,26 @@ public void close() {
     // NOTE: DO NOT close the PinotDataBuffer here because it is tracked by the caller and might be reused later. The
     // caller is responsible of closing the PinotDataBuffer.
   }
+
+  private ImmutableRoaringBitmap getMatchesInRange(int firstRangeId, int lastRangeId) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/RangeIndexReaderImpl.java:276]: Good to add some comments on how the -1 is handled (list the different scenarios)

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/RangeIndexReaderImpl.java [hunk 6]
```diff
@@ -217,4 +272,26 @@ public void close() {
...
+    if (firstRangeId == lastRangeId) {
+      return null;
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/RangeIndexReaderImpl.java:278]: If both first and last is -1, it might be no match or full match. We need to handle them separately

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java [hunk 7]
```diff
@@ -51,74 +51,84 @@ public RangeIndexBasedFilterOperator(PredicateEvaluator rangePredicateEvaluator,
...
     int firstRangeId;
     int lastRangeId;
     if (_rangePredicateEvaluator instanceof SortedDictionaryBasedRangePredicateEvaluator) {
-      firstRangeId = rangeIndexReader
-          .findRangeId(((SortedDictionaryBasedRangePredicateEvaluator) _rangePredicateEvaluator).getStartDictId());
       // NOTE: End dictionary id is exclusive in OfflineDictionaryBasedRangePredicateEvaluator.
-      lastRangeId = rangeIndexReader
-          .findRangeId(((SortedDictionaryBasedRangePredicateEvaluator) _rangePredicateEvaluator).getEndDictId() - 1);
+      matches = rangeIndexReader.getMatchingDocIds(
+          ((SortedDictionaryBasedRangePredicateEvaluator) _rangePredicateEvaluator).getStartDictId(),
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java:69]: (nit) Cache the `min` and `max` value

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java [hunk 8]
```diff
@@ -51,74 +51,84 @@ public RangeIndexBasedFilterOperator(PredicateEvaluator rangePredicateEvaluator,
...
       }
     }
-
-    // Need to scan the first and last range as they might be partially matched
-    // TODO: Detect fully matched first and last range
-    ImmutableRoaringBitmap docIdsToScan;
-    if (firstRangeId == lastRangeId) {
-      docIdsToScan = rangeIndexReader.getDocIds(firstRangeId);
+    // this branch is likely until RangeIndexReader reimplemented and enabled by default
+    if (null != partialMatches) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java:113]: (nit) we usually put `(partialMatches != null)`, same for other places

**Review by Jackie-Jiang — COMMENTED** [review]
I feel the current raw value based range index does not work when the range id is `-1`, or I might miss something. Can we first check if the current one is correct before this refactoring?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/RangeIndexReaderImpl.java (hunk 4 above)
Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/RangeIndexReaderImpl.java:281]: No, this is correct. `firstRangeId == 0` means that the lower bound was below the lowest bucket, so range 0 (`firstRangeId + 1`) is within the range of the query so long as `lastRangeId > 0`. Similarly, a `lastRangeId == -1` means that the upper bound of the query was below the first range, which means that none of the rows are within the range of the query, so an empty bitmap is produced.

Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/RangeIndexReaderImpl.java:281]: Based on my reading, the `findRangeId()` will return `-1` if the value is smaller than the lower bound of the lowest bucket or larger than the higher bound of the highest bucket. What if the `lastRangeId == -1` because it is larger than the higher bound?

Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/RangeIndexReaderImpl.java:281]: Actually, Looking back at `findRangeId`, it is ambiguous. -1 can mean smaller than the lowest range or greater than the highest range. So this was always broken.

This code was lifted straight from `RangeIndexBasedFilterOperator`:
```java
// Ranges in the middle of first and last range are fully matched
     for (int rangeId = firstRangeId + 1; rangeId < lastRangeId; rangeId++) {
       docIds.or(rangeIndexReader.getDocIds(rangeId));
     }
```

I don't want existing bugs/specification flaws to get in the way of redefining this functionality.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/RangeIndexReaderImpl.java (hunk 6 above)
Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/RangeIndexReaderImpl.java:278]: Not necessarily, -1 currently means both smaller than the smallest range and larger than the largest range, so it could mean a full match or no match depending on the original query parameters. This is consistent with the current behaviour though, which appears to be incorrect.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java (hunk 7 above)
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java:69]: Why?

**Comment by richardstartin** [comment]
The way I see it, this feature is buggy:
* it converted floats to ints 
* it doesn't produce evenly sized ranges
* it doesn't handle or test rangeIds of -1, which are ambiguous and can mean smaller than the smallest range _or_ larger than the largest range, and `RangeIndexReader.getDocIds` never knew the difference.

If this feature is in use, these bugs are probably acceptable to users. What I'm trying to do here is move some code around so I can eventually replace the feature, I don't want to fix every problem with the existing implementation in the process of doing so.

**Comment by richardstartin** [comment]
I will apply the requested code style changes, but let's look into setting up check style rules to enforce them in another PR, as it's a waste of everybody's time to enforce style guidelines manually during code review.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java (hunk 7 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java:69]: Both performance and readability. What I mean is to put `((SortedDictionaryBasedRangePredicateEvaluator) _rangePredicateEvaluator).getStartDictId()` and `((SortedDictionaryBasedRangePredicateEvaluator) _rangePredicateEvaluator).getEndDictId() - 1` into local variables instead of calculating them twice

**Comment by Jackie-Jiang** [comment]
> The way I see it, this feature is buggy:
> 
> * it converted floats to ints
> * it doesn't produce evenly sized ranges
> * it doesn't handle or test rangeIds of -1, which are ambiguous and can mean smaller than the smallest range _or_ larger than the largest range, and `RangeIndexReader.getDocIds` never knew the difference.
> 
> If this feature is in use, these bugs are probably acceptable to users. What I'm trying to do here is move some code around so I can eventually replace the feature, I don't want to fix every problem with the existing implementation in the process of doing so.

I don't agree that we should leave the bug as is just because no one complains. It is hard for user to realize the result returned is inaccurate.
It is okay to not fix it in the refactoring PR, but we should at least add a TODO and create an issue to track it so that it can be fixed in a following PR.
Also, it shouldn't be hard to fix:
- Convert float to int - already fixed in this PR
- Not even sized range - not optimal but won't cause wrong result
- RangeId of -1 - We can change the upper bound id to Integer.MAX_VALUE to differentiate them and handle them properly

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java (hunk 7 above)
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java:69]: There is no performance benefit to the suggested change, these values are constants in `SortedDictionaryBasedRangePredicateEvaluator` (and all other implementations used in this block of code):

```java
    public int getStartDictId() {
      return _startDictId;
    }

    public int getEndDictId() {
      return _endDictId;
    }
```

I will make the change on the basis of readability, but it's very difficult to spot bottlenecks by code review so it's better to rely on data.

**Comment by richardstartin** [comment]
I've fixed the pre-existing bugs and have added tests for these cases, which fail on master.

**Comment by richardstartin** [comment]
Looks like flaky tests are being flaky again, please rerun CI.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/RangeIndexReader.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+  @Nullable
+  T getMatchingDocIds(double min, double max);
+
+  /**
+   * Returns doc ids with a value between min and max, both inclusive.
+   * @param min the inclusive lower bound.
+   * @param max the inclusive upper bound.
+   * @return the matching doc ids.
+   */
+  @Nullable
```
Comment (klsince) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/RangeIndexReader.java:71]: looks like the comments are identical for getMatchingDocIds() and getPartiallyMatchingDocIds(). I assume their behavior would be different. If so, perhaps highlight it in the comments?

**Review by Jackie-Jiang — COMMENTED** [review]
Implementation looks good. The test can be flaky if we remove the random seed.
One way to simplify the test is to explicitly check the range id of the extreme values. To catch the typo of storing float as int, we can check whether the range start array values are all existing values

File: pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/RangeIndexCreatorTest.java [hunk 10]
```diff
@@ -35,11 +37,12 @@
 import org.testng.annotations.Test;
 
 import static org.apache.pinot.segment.spi.V1Constants.Indexes.BITMAP_RANGE_INDEX_FILE_EXTENSION;
+import static org.testng.Assert.*;
 
 
 public class RangeIndexCreatorTest {
   private static final File INDEX_DIR = new File(FileUtils.getTempDirectory(), "RangeIndexCreatorTest");
-  private static final Random RANDOM = new Random();
+  private static final Random RANDOM = new Random(42);
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/RangeIndexCreatorTest.java:45]: tests should be deterministic

File: pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/RangeIndexCreatorTest.java [hunk 11]
```diff
@@ -106,23 +109,22 @@ private void testDataType(DataType dataType)
...
     int numDocs = 1000;
-    Number[] values = new Number[numDocs];
+    Object values = valuesArray(dataType, numDocs);
 
+    int numValuesPerRange;
     try (RangeIndexCreator creator = new RangeIndexCreator(INDEX_DIR, fieldSpec, dataType, -1, -1, numDocs, numDocs)) {
       addDataToIndexer(dataType, numDocs, 1, creator, values);
       creator.seal();
+      // account for off by one bug in v1 implementation
+      numValuesPerRange = creator.getNumValuesPerRange() + 1;
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/RangeIndexCreatorTest.java:119]: The current implementation reports 50 values per bucket but assigns 51 values to the first 19 buckets and 31 to the last one.

File: pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/RangeIndexCreatorTest.java [hunk 12]
```diff
@@ -242,141 +242,352 @@ private void addDataToIndexer(DataType dataType, int numDocs, int numValuesPerEn
...
+        assertEquals(matches, rangeIndexReader.getPartiallyMatchingDocIds(coveredRange[0], coveredRange[1]));
+        // partial matches must be the combination of the two edge buckets
+        ImmutableRoaringBitmap partialMatches = rangeIndexReader.getPartiallyMatchingDocIds(
+            lowerPartialRange[0], upperPartialRange[1]);
+        assertNotNull(partialMatches, "partial matches for single range must not be null");
+        assertEquals(ImmutableRoaringBitmap.or(
+            rangeIndexReader.getPartiallyMatchingDocIds(lowerPartialRange[0], lowerPartialRange[1]),
+            rangeIndexReader.getPartiallyMatchingDocIds(upperPartialRange[0], upperPartialRange[1])), partialMatches);
+        // edge cases
+        assertEquals(((int[]) values).length, numValuesPerMVEntry
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/RangeIndexCreatorTest.java:279]: This API doesn't exist for `Random.nextLong()`. The `Random` is seeded for determinism, as is good practice in testing (this is not a fuzz test, this is not property based testing) so this is not a problem.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/RangeIndexReader.java (hunk 2 above)
Comment (richardstartin) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/RangeIndexReader.java:28]: I don't consider this important at this stage.

File: pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/RangeIndexCreatorTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (richardstartin) [reply pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/RangeIndexCreatorTest.java:45]: https://martinfowler.com/articles/nonDeterminism.html

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/RangeIndexReader.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/RangeIndexReader.java:28]: Sure that is fine. I saw you replied with thumb up and thought you want to change that

File: pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/RangeIndexCreatorTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/RangeIndexCreatorTest.java:119]: I see, you are right. The `if (i > start + boundary)` in the current implementation is causing this, which should be `if (i >= start + boundary)` instead.
Since we're going to deprecate the old implementation, okay to leave it as is

File: pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/RangeIndexCreatorTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/RangeIndexCreatorTest.java:45]: I think the non-determinism in the article doesn't refer to fixing the input data, but the test should not be flaky (in the article: `In talking with development teams I've often heard about the problem of non-deterministic tests - tests that sometimes pass and sometimes fail.`).
The reason why we use random input values is because we usually cannot come up with (or too time consuming to create) all different inputs to cover all corner cases, so we use random input to do incremental checks during each CI run. The test should pass regardless of the input values, and when we detect test flakiness, we can check and fix the implementation (or the test).
I think this test can be flaky in certain cases (e.g. MIN/MAX integer values are generated, same values are generated and split into different ranges), but since the old implementation is going to be deprecated, we can leave it as is and not spend time fixing this test.

**Review by Jackie-Jiang — APPROVED** [review]
LGTM. Thanks for fixing the bugs in the current range index reader. Will merge once the commit of resolved conversation is pushed.
